File: unicode-hs/src/Unicode/Codec/Bom.hs
{-|
Module      : Unicode.Codec.Bom
Description : Byte-Order-Mark detection across the five Unicode encodings.

Haskell port of @Unicode.Codec.Bom@ from unicode-lean.

@
UTF-8    : EF BB BF              (3 bytes)
UTF-16BE : FE FF                 (2 bytes)
UTF-16LE : FF FE                 (2 bytes)
UTF-32BE : 00 00 FE FF           (4 bytes)
UTF-32LE : FF FE 00 00           (4 bytes)
@

Order matters: UTF-32 BOMs share their leading bytes with UTF-16 BOMs,
so the 4-byte UTF-32 patterns must be checked BEFORE the 2-byte UTF-16
patterns. Specifically, @FF FE 00 00@ is a UTF-32LE BOM, not a UTF-16LE
BOM followed by two NUL bytes.

Constructor names match the Lean @BomKind.utf8@ / @BomKind.utf16BE@ etc.
(PascalCase is language-mandated in Haskell). Consume via qualified
import:

> import qualified Unicode.Codec.Bom as Bom
> case Bom.detect bs of
>   Just (Bom.Utf8, _) -> ...
-}
module Unicode.Codec.Bom
  ( BomKind (..)
  , bomLength
  , detect
  , strip
  ) where

import Data.ByteString (ByteString)
import qualified Data.ByteString as BS
import Data.Word (Word8)

-- | The five Unicode encoding kinds distinguishable by their BOM.
data BomKind
  = Utf8
  | Utf16BE
  | Utf16LE
  | Utf32BE
  | Utf32LE
  deriving stock (Eq, Show, Ord, Enum, Bounded)

-- | The byte length of each BOM.
bomLength :: BomKind -> Int
bomLength Utf8    = 3
bomLength Utf16BE = 2
bomLength Utf16LE = 2
bomLength Utf32BE = 4
bomLength Utf32LE = 4

-- | Detect a leading BOM, returning the encoding kind and the number of
-- BOM bytes to skip. The 4-byte UTF-32 BOMs are tested before the 2-byte
-- UTF-16 BOMs because @FF FE 00 00@ is UTF-32LE, not UTF-16LE followed
-- by U+0000. Returns 'Nothing' if the input does not begin with any
-- recognised BOM.
detect :: ByteString -> Maybe (BomKind, Int)
detect bs
  | n >= 4
  , b0 == 0x00, b1 == 0x00, b2 == 0xFE, b3 == 0xFF = Just (Utf32BE, 4)
  | n >= 4
  , b0 == 0xFF, b1 == 0xFE, b2 == 0x00, b3 == 0x00 = Just (Utf32LE, 4)
  | n >= 3
  , b0 == 0xEF, b1 == 0xBB, b2 == 0xBF             = Just (Utf8, 3)
  | n >= 2
  , b0 == 0xFE, b1 == 0xFF                         = Just (Utf16BE, 2)
  | n >= 2
  , b0 == 0xFF, b1 == 0xFE                         = Just (Utf16LE, 2)
  | otherwise                                      = Nothing
  where
    !n = BS.length bs
    byteAt :: Int -> Word8
    byteAt i = if i < n then BS.index bs i else 0
    !b0 = byteAt 0
    !b1 = byteAt 1
    !b2 = byteAt 2
    !b3 = byteAt 3

-- | Strip the BOM from @bs@ if one is present, returning the remaining
-- content and the detected encoding. Returns @(Nothing, bs)@ (no BOM
-- stripped) if the input does not begin with a recognised BOM.
strip :: ByteString -> (Maybe BomKind, ByteString)
strip bs = case detect bs of
  Just (kind, k) -> (Just kind, BS.drop k bs)
  Nothing        -> (Nothing, bs)

File: src/unicode_python/bom.py
"""Byte-Order-Mark detection across the five Unicode encodings.

    UTF-8     : EF BB BF              (3 bytes)
    UTF-16 BE : FE FF                 (2 bytes)
    UTF-16 LE : FF FE                 (2 bytes)
    UTF-32 BE : 00 00 FE FF           (4 bytes)
    UTF-32 LE : FF FE 00 00           (4 bytes)

Order matters: the UTF-32 BOMs share their leading bytes with the
UTF-16 BOMs, so the 4-byte patterns must be checked BEFORE the
2-byte patterns.  Specifically, ``FF FE 00 00`` is a UTF-32 LE
BOM, not a UTF-16 LE BOM followed by two NUL bytes.
"""

from enum import Enum


class BomKind(Enum):
    """The five Unicode encoding kinds distinguishable by their BOM."""

    UTF8 = "Utf8"
    UTF16_BE = "Utf16BE"
    UTF16_LE = "Utf16LE"
    UTF32_BE = "Utf32BE"
    UTF32_LE = "Utf32LE"


_LENGTHS: dict[BomKind, int] = {
    BomKind.UTF8: 3,
    BomKind.UTF16_BE: 2,
    BomKind.UTF16_LE: 2,
    BomKind.UTF32_BE: 4,
    BomKind.UTF32_LE: 4,
}


def bom_length(kind: BomKind) -> int:
    """The byte length of each BOM."""
    return _LENGTHS[kind]


def _byte_at(data: bytes, i: int) -> int:
    return data[i] if i < len(data) else 0


def detect(data: bytes) -> tuple[BomKind, int] | None:
    """Detect a leading BOM, returning the encoding kind and the
    number of BOM bytes to skip.

    The 4-byte UTF-32 BOMs are tested before the 2-byte UTF-16
    BOMs.  Returns ``None`` if the input does not begin with any
    recognised BOM.
    """
    b0 = _byte_at(data, 0)
    b1 = _byte_at(data, 1)
    b2 = _byte_at(data, 2)
    b3 = _byte_at(data, 3)
    if len(data) >= 4 and b0 == 0x00 and b1 == 0x00 and b2 == 0xFE and b3 == 0xFF:
        return (BomKind.UTF32_BE, 4)
    if len(data) >= 4 and b0 == 0xFF and b1 == 0xFE and b2 == 0x00 and b3 == 0x00:
        return (BomKind.UTF32_LE, 4)
    if len(data) >= 3 and b0 == 0xEF and b1 == 0xBB and b2 == 0xBF:
        return (BomKind.UTF8, 3)
    if len(data) >= 2 and b0 == 0xFE and b1 == 0xFF:
        return (BomKind.UTF16_BE, 2)
    if len(data) >= 2 and b0 == 0xFF and b1 == 0xFE:
        return (BomKind.UTF16_LE, 2)
    return None


def strip(data: bytes) -> tuple[BomKind | None, bytes]:
    """Strip the BOM from ``data`` if one is present, returning the
    remaining content together with the detected encoding.  When
    the input does not begin with a recognised BOM the function
    returns ``(None, data)``.
    """
    result = detect(data)
    if result is None:
        return (None, data)
    kind, k = result
    return (kind, data[k:])


__all__ = ["BomKind", "bom_length", "detect", "strip"]
